packages/context/src/binding.ts

```diff
@@ -92,6 +96,15 @@ export enum BindingType {
   PROVIDER = 'Provider',
 }
 
+export type BindingOptions =
+  | {[name: string]: BoundValue}
+  | Promise<{[name: string]: BoundValue}>;
+
+type GetValueFunction<T> = (
+  ctx?: Context,
+  session?: ResolutionSession,
+) => T | PromiseLike<T | RejectionError>;
```

Comment:
IIUC, `GetValueFunction` has the following possible outcomes (return values)

 1. a sync value `T`
 2. an error was thrown synchronously
 3. a resolved promise with a sync value `T`
 4. a resolved promise with a rejection error
 5. a rejected promise

I find it confusing that we have two ways how to signal an error - either via resolving a promise with `RejectionError` or by rejecting the promise. 

I am also concerned that we can accidentally leak promises resolved to a rejection error as a successfully resolved value, when somebody contributes new code calling `GetValueFunction` and does not realize that it may return a resolved promise even on an error.

@raymondfeng What is the problem you are trying to solve with this approach? What are the downsides of reporting RejectionError as a regular error through promise rejection?

If we decided to keep the current design (I hope not), then shouldn't we allow `GetValueFunction` to return `RejectionError` synchronously too?

```ts
T | RejectionError | PromiseLike<T | RejectionError>
// even better:
ValueOrPromise<T | RejectionError>
```

Reply:
I think I am starting to understand the problem you are trying to solve:

 - The application setup code sets binding options to a promise. This promise is stored for the time being.
 - Only later, when the binding is resolved, the promise is awaited and errors are detected.

I see a usability problem here: if the async code producing binding options fails, then we have an application-setup error. Ideally, this app-setup error should be caught and reported as soon as possible, and it should abort application startup. Right now, the error is silently stored and manifests only if somebody (e.g. a REST client) triggers the execution path that requires the particular binding with faulty options to be resolved. In my experience that may lead to unnecessary outages that are difficult to troubleshoot, as the error may be triggered days or even months after the new application version was deployed, at which time nobody will remember the context in which the faulty options were introduced. (To make things even worse, in Continuous Deployment mode, there may be several new versions deployed after the problem was introduced but before it manifests.) I think we should really look for a way how to report faulty options right when the application starts.

Setting the usability aside, if we decide to keep the current approach, then I am proposing to use a slightly different design here, which I consider to be most robust and also compatible with `async`/`await`.

Instead of storing async-computed options as a promise that may be resolved to a RejectionError, I am proposing to store them as an async getter function. Instead of converting a rejected promise into a promise of RejectionError, we will convert it into a function returning a rejected promise. (Similarly, resolved promises will be converted to a function returning a resolved value.) The example below shows my proposed version of `RejectionError.catch`, I think the code should be moved to a standalone helper function called e.g. `delayPromiseRejection`.

```ts
class RejectionError {
  static catch<T>(p: ValueOrPromise<T>): Getter<ValueOrPromise<T>> {
    if (p instanceof Promise) {
      return p.then(
        result => () => result,
        error => () => Promise.reject(error));
    } else {
      return () => p;
   }
}

// Hopefully this will be no longer needed, but
// I am posting it here to better illustrate my proposal
type GetValueFunction<T> = (
  ctx?: Context,
  session?: ResolutionSession,
) => Getter<ValueOrPromise<T>>
```

Instead of calling `RejectionError.then`, we can simply invoke the getter function inside `try`/`catch`:

```ts
const resolutionSession = ResolutionSession.enterBinding(this, session);
try {
  const val = await this._getValue(ctx, resolutionSession)();
  resolutionSession.exit();
  return val;
} catch (err) { 
  // handle errors if needed
}
```

Thoughts?

Reply:
Unfortunately Node.js will complain if a promise does not have a `catch` handler set up before the rejection.  Propagating the rejected promise and trying to handle the error later does not help. The warning is similar as follows:
```
(node:7741) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: woops
(node:7741) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

That's why I had to introduce `RejectionError` to transform a rejected promise into a normal promise that resolves to an instance of `RejectionError` so that the error can be gracefully bubbled up to the client and only return a rejected promise at last.

Reply:
@raymondfeng I understand what you are saying, but it seems to me you don't understand my proposal. Sorry for not being clear enough!

In my proposal, I am attaching the `catch` handler on the same tick of the event loop, explicitly storing the promise rejection (so that Node.js does not complain) and then raising the error again only after the getter function has been called.

Let's start with your example:

```js
const p = Promise.reject("error 1");

setTimeout(() => {
  // It's too later to catch the error from the rejected promise
  p.catch(err => console.log(err));
}, 100);
```

and rewrite it show what I mean:

```js
const p = Promise.reject("error 1");
const storedPromise = p.then(
  result => () => result,
  error => () => Promise.reject(error));

setTimeout(() => {
  // The rejection promise was silently stored away from V8/Node.js eyes
  storedPromise.then(getter => getter().catch(err => console.log(err)));
}, 100);
```

When I run my version, Node.js prints the following message:

```
$ node test.js
error 1
```

In the particular case of binding options, it means that internally, `Binding.withOptions` stores either the options object itself, or else a promise that will be resolved to a getter function returning a promise of the options object. In my proposal, the conversion from a promise of the options object (provided by the user - passed as an argument to `withOptions`) to a promise of a getter function (think about `storedPromise` in the simplified example code posted earlier in this comment) should be implemented inside `withOptions` as an implementation detail of that method.

Having said that, there is still the usability issue of deferring promise rejections from `withOptions` - see my https://github.com/strongloop/loopback-next/pull/671#issuecomment-355984034. If we agree that promise rejections related to `withOptions` setup should be reported immediately, then this entire discussion is moot.

Reply:
The introduction of `RejectionError` is to allow propagation of rejected promises (which can be a `throw` inside an async function) for the dependency injection chain. It's NOT particularly for `withOptions`. The wrapping/unwrapping of rejected promises cannot stay only as internal for `withOptions`. I use `chain` to refer to the stack of injections, such as `A[arg 0] -> B[prop c] -> C[arg 1] -> D`. Any step of the resolution can generate rejected promises. I was hoping async/await does the graceful error bubbling but we know that is not the case.

I think now we both agree that a rejected promise needs to be wrapped with a `catch` handler immediately to prevent Node from complaining about the unhandled promise rejection. The only difference is that how it is wrapped. My proposal was to wrap the rejected promise into an instance of `RejectionError` and leave the normal value as is and yours is to wrap them into a closure/function. I think both of them work but the complexity of propagating the wrapper through the whole chain is the same.

Reply:
> Any step of the resolution can generate rejected promises. I was hoping async/await does the graceful error bubbling but we know that is not the case.

This does not match my understanding of how promises and async/await works in JavaScript. Perhaps I am misunderstanding what you are suggesting? Could you please show me an example code demonstrating how a regular chain of injections `A[arg 0] -> B[prop c] -> C[arg 1] -> D` produces an "unhandled promise rejection" warning?

I understand why this becomes a problem in `withOptions` scenario. What I don't understand is how and why this triggers the unhandled promise rejection warning in the regular case.

> The introduction of RejectionError is to allow propagation of rejected promises (which can be a throw inside an async function) for the dependency injection chain. It's NOT particularly for withOptions. The wrapping/unwrapping of rejected promises cannot stay only as internal for withOptions
> I think now we both agree that a rejected promise needs to be wrapped with a catch handler immediately to prevent Node from complaining about the unhandled promise rejection. 

I agree that a rejected promise needs to be wrapped with a catch handler in the same tick of the event loop.

According to my understanding of JavaScript, this does not imply that we have to change all code paths to handle RejectionError now. In my mind, it's only `withOptions` feature that receives a promise argument in one tick of event loop and attach then/catch handlers to it only much later. Other code paths (e.g. calling `ctx.get` to create a class with dependencies) should not be suffering from the problem as long as the caller of `ctx.get` attaches `then` or `catch` callback immediately, or uses `await cxt.get` (preferably).

@raymondfeng I have a proposal for you to consider. What if we left support for async options (`withOptions` receiving a promise instead of the actual options object) out of scope of this pull request, so that we can get _something_ landed soon, and then have more space to look into this async/promise flow control to find a good solution?

Reply:
This simple example demonstrates the issue:
```js
// Simulate how we resolve two arguments of a constructor asynchronously
async function a() {
  return [
    /*
    new Promise((resolve, reject) => {
      setImmediate(() => {
        reject(new Error("error"));
      });
    }),
    */
    Promise.reject(new Error("error")), // 1st arg
    Promise.resolve("arg2") // 2nd arg
  ];
}

// Join the result of all args and call the ctor
function b() {
  return Promise.all(a());
}

// Try-catch 
async function c() {
  try {
    await b();
  } catch (e) {
    console.error(e);
  }
}

c();
```

Reply:
@bajtos Check this out - https://github.com/strongloop/loopback-next/commit/ed7c5a5468f42970be3ec15d42d7fa6576ffb5da. Run `npm test` and you'll see the unhandled promise rejections.

Reply:
https://jsblog.insiderattack.net/promises-next-ticks-and-immediates-nodejs-event-loop-part-3-9226cbe7a6aa

Reply:
Now it becomes clear why https://github.com/strongloop/loopback-next/commit/ed7c5a5468f42970be3ec15d42d7fa6576ffb5da reports unhandled promise rejection. Mocha uses `setImmediate` (instead of `process.nextTick`) to switch control from `before` to `it`. See https://github.com/mochajs/mocha/issues/3009. As a result, the rejected promise in `before` is too late for the `catch` inside `it`.

Reply:
@raymondfeng first of all, thank you for creating the example test in ed7c5a5, it helped me a lot to better understand what problem(s) you are trying to solve here.

I was thinking about this problem quite a lot in the past two days, even before reading your example.

IMO, the problems we are facing with unhandled rejection warnings are telling us that we are using Promises in a way they were not designed for. Promises are a construct primarily intended for flow control (in algorithm with steps 1 and 2, we want to wait for the outcome of step 1 before starting step 2), not a tool for storing values that may get available only in the future, depending on the outcome of a background async task (as you are doing in `withOptions` and the example in ed7c5a5).

IMO, both `withOptions` and ed7c5a5 are exhibiting the same design flaw: a flow control primitive (a Promise that will be resolved or rejected in the future) is stored as a static-like value that is later retrieved for use.

In particular, I consider `ctx.bind('foo').to(new Promise(/*...*/))` as an anti-pattern. Dynamically computed values should be bound via a getter function and `.toDynamicValue` method.

```ts
// WRONG
    ctx.bind('foo').to(
      new Promise((resolve, reject) => {
        setImmediate(() => {
          reject(new Error('foo: error'));
        });
      }),
    );

// GOOD
    ctx.bind('foo').toDynamicValue(
      () =>
        new Promise((resolve, reject) => {
          setImmediate(() => {
            reject(new Error('foo: error'));
          });
        }),
    );
```

I pushed 7c53fe8 to your feature branch to prove my point - the test is passing now with no warnings printed.

In the light of the above, I am proposing to refactor `withOptions` signature to accept either `options` value (the object) or a getter function that will be used to obtain the value (the object) in the future. We can allow this getter function to be both sync or async (to support the use case where binding options require async computation), and in the future we can memoize the returned value (promise resolution) as a performance optimization (if needed at all).


My proposal will also allow us to compute options lazily, only at the time they are required for the first time. I think this makes the error handling more intuitive. Because we execute the code building binding option only at options' first use, it's only natural that we also report errors from this async computation only at the time the options are actually used by a binding. Compare it with your original proposal, where the async computation is started before calling `withOptions`, but the outcome (the error) is reported only later when the options are accessed.

Reply:
I opened a pull request to modify `Binding#to` to reject promise values - see https://github.com/strongloop/loopback-next/pull/854

Reply:
The discussion has driven great understanding of how promises works and how to use them. I think it will be fantastic topic for a blog.